File: library/src/main/java/io/ashdavies/rx/rxfirebase/RxFirebaseAuth.java
package io.ashdavies.rx.rxfirebase;

import android.support.annotation.CheckResult;

import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.ProviderQueryResult;

import io.ashdavies.rx.common.NullableMaybeOnSubscribe;
import io.ashdavies.rx.rxtasks.RxTasks;
import io.reactivex.BackpressureStrategy;
import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Maybe;
import io.reactivex.Single;

@SuppressWarnings("WeakerAccess")
public final class RxFirebaseAuth {
  private final FirebaseAuth auth;

  private RxFirebaseAuth(FirebaseAuth auth) {
    this.auth = auth;
  }

  public static RxFirebaseAuth getInstance() {
    return getInstance(FirebaseAuth.getInstance());
  }

  public static RxFirebaseAuth getInstance(FirebaseAuth firebaseAuth) {
    return new RxFirebaseAuth(firebaseAuth);
  }

  @CheckResult
  public Flowable<FirebaseAuth> authStateChanged() {
    return Flowable.create(new AuthStateOnSubscribe(auth), BackpressureStrategy.BUFFER);
  }

  @CheckResult
  public Single<AuthResult> signInWithCredential(AuthCredential authCredential) {
    return RxTasks.single(auth.signInWithCredential(authCredential));
  }

  @CheckResult
  public Single<AuthResult> signInWithCustomToken(String token) {
    return RxTasks.single(auth.signInWithCustomToken(token));
  }

  @CheckResult
  public Single<AuthResult> signInWithEmailAndPassword(String email, String password) {
    return RxTasks.single(auth.signInWithEmailAndPassword(email, password));
  }

  @CheckResult
  public Single<AuthResult> signInAnonymously() {
    return RxTasks.single(auth.signInAnonymously());
  }

  @CheckResult
  public Single<AuthResult> createUserWithEmailAndPassword(String email, String password) {
    return RxTasks.single(auth.createUserWithEmailAndPassword(email, password));
  }

  @CheckResult
  public Single<ProviderQueryResult> fetchProvidersForEmail(String email) {
    return RxTasks.single(auth.fetchProvidersForEmail(email));
  }

  @CheckResult
  public Maybe<FirebaseUser> getCurrentUser() {
    return Maybe.create(new NullableMaybeOnSubscribe<>(auth.getCurrentUser()));
  }

  @CheckResult
  public Completable sendPasswordResetEmail(String email) {
    return RxTasks.completable(auth.sendPasswordResetEmail(email));
  }

  public RxFirebaseAuth signOut() {
    auth.signOut();
    return this;
  }
}

File: src/main/kotlin/io/ashdavies/rx/rxfirebase/RxFirebaseAuth.kt
package io.ashdavies.rx.rxfirebase

import androidx.annotation.CheckResult
import com.google.firebase.auth.AuthCredential
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.ProviderQueryResult
import com.google.firebase.auth.SignInMethodQueryResult
import io.ashdavies.rx.extensions.onAuthState
import io.ashdavies.rx.rxtasks.toCompletable
import io.ashdavies.rx.rxtasks.toSingle
import io.reactivex.BackpressureStrategy
import io.reactivex.BackpressureStrategy.BUFFER
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Maybe
import io.reactivex.Single

@Deprecated("Use RxTasks instead")
data class RxFirebaseAuth(private val auth: FirebaseAuth = FirebaseAuth.getInstance()) {

  @CheckResult
  fun currentUser(): Maybe<FirebaseUser> = Maybe.fromCallable { auth.currentUser }

  @CheckResult
  @JvmOverloads
  fun authStateChanged(strategy: BackpressureStrategy = BUFFER): Flowable<FirebaseAuth> = auth
      .onAuthState()
      .toFlowable(strategy)

  @CheckResult
  fun signInWithCredential(credential: AuthCredential): Single<AuthResult> = auth
      .signInWithCredential(credential)
      .toSingle()

  @CheckResult
  fun signInWithCustomToken(token: String): Single<AuthResult> = auth
      .signInWithCustomToken(token)
      .toSingle()

  @CheckResult
  fun signInWithEmailAndPassword(email: String, password: String): Single<AuthResult> = auth
      .signInWithEmailAndPassword(email, password)
      .toSingle()

  @CheckResult
  fun signInAnonymously(): Single<AuthResult> = auth
      .signInAnonymously()
      .toSingle()

  @CheckResult
  fun createUserWithEmailAndPassword(email: String, password: String): Single<AuthResult> = auth
      .createUserWithEmailAndPassword(email, password)
      .toSingle()

  @CheckResult
  @Deprecated("Use fetchSignInMethodsForEmail instead")
  fun fetchProvidersForEmail(email: String): Single<ProviderQueryResult> = auth
      .fetchProvidersForEmail(email)
      .toSingle()

  @CheckResult
  fun fetchSignInMethodsForEmail(email: String): Single<SignInMethodQueryResult> = auth
      .fetchSignInMethodsForEmail(email)
      .toSingle()

  @CheckResult
  fun sendPasswordResetEmail(email: String): Completable = auth
      .sendPasswordResetEmail(email)
      .toCompletable()

  @CheckResult
  fun signOut(): RxFirebaseAuth = apply { auth.signOut() }

  companion object {

    @JvmStatic
    fun getInstance() = RxFirebaseAuth()

    @JvmStatic
    fun getInstance(auth: FirebaseAuth) = RxFirebaseAuth(auth)
  }
}
